### PR TITLE
Add basic implementation of nats client

### PR DESCRIPTION
**Pull request by mgrachev**

### CODE REVIEW

File: src/headers.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+pub struct HeaderMapBuilder {
+    inner: async_nats::HeaderMap,
+}
+
+impl HeaderMapBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn nats_message_id(mut self, message_id: &str) -> Self {
```
Comment (manifest) [inline src/headers.rs:18]: Maybe just a `message_id` ?

**Review by 0nkery — APPROVED** [review]

File: src/client.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+pub async fn new(url: &str, creds: &str) -> Result<Client> {
+    let client = async_nats::ConnectOptions::with_credentials_file(creds.into())
+        .await?
+        .event_callback(|event| async move {
+            match event {
+                Event::ServerError(error) => {
+                    error!(%error, "server error occurred");
+                }
+                Event::ClientError(error) => {
+                    error!(%error, "server error occurred");
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/client.rs]: Fixed

File: src/headers.rs (hunk 1 above)
Comment (mgrachev) [reply src/headers.rs]: Done

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#[async_trait]
+pub trait NatsClient: Send + Sync {
+    async fn publish(
+        &self,
+        subject: String,
+        payload: Vec<u8>,
+        headers: Option<HeaderMap>,
+    ) -> Result<()>;
+
+    async fn pull_messages(
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/lib.rs]: Done

File: src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        subject: String,
+        payload: Vec<u8>,
+        headers: Option<HeaderMap>,
+    ) -> Result<()> {
+        self.jetstream
+            .publish_with_headers(subject, headers.unwrap_or_default().inner, payload.into())
+            .await
+            .map_err(|e| anyhow!("failed to publish message through nats: {}", e))?
+            .await
+            .map_err(|e| anyhow!("failed to ack message: {}", e))?;
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/lib.rs]: Replaced with `thiserror`

File: src/client.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SubscribeError {
+    #[error("failed to get stream")]
+    GettingStreamFailed(String),
+    #[error("failed to get consumer")]
+    GettingConsumerFailed(String),
+    #[error("failed to create stream of messages")]
+    StreamCreationFailed(String),
```
Comment (manifest) [inline src/client.rs:54]: We have some details for each error kind, but don't serialize those details. What the use of them?

File: src/headers.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                    async_nats::header::NATS_MESSAGE_ID.to_string(),
+                ))?;
+        let event_id = event_id.try_into().map_err(HeaderError::InvalidEventId)?;
+
+        let sender_id = value
+            .get(async_nats::header::NATS_MESSAGE_ID)
+            .ok_or(HeaderError::InvalidHeader(SENDER_AGENT_ID.to_string()))?
+            .as_str();
+        let sender_id = AgentId::from_str(sender_id)
+            .map_err(|e| HeaderError::SenderIdParseFailed(e.to_string()))?;
```
Comment (manifest) [inline src/headers.rs:76]: Do we need to convert an internal error type to string?

File: src/client.rs (hunk 5 above)
Comment (manifest) [reply src/client.rs:53]: Same here. We convert rich types to strings.

File: src/headers.rs (hunk 6 above)
Comment (mgrachev) [reply src/headers.rs]: No, fixed.

File: src/client.rs (hunk 5 above)
Comment (mgrachev) [reply src/client.rs:53]: `async_nats` returns errors as `Box<dyn std::error::Error + Send + Sync>`, so I convert them to strings.

File: src/client.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+};
+use std::str::FromStr;
+use tracing::{error, warn};
+
+#[derive(Clone)]
+pub struct Client {
+    jetstream: Context,
+}
+
+pub async fn new(url: &str, creds: &str) -> Result<Client, ConnectError> {
```
Comment (0nkery) [inline src/client.rs:19]: Why not `Client::new`?

File: src/config.rs [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+use std::time::Duration;
+
+#[derive(Clone, Debug, Deserialize)]
+pub struct Config {
+    pub url: String,
+    pub creds: String,
+    pub stream: Option<String>,
+    pub consumer: Option<String>,
+    #[serde(with = "humantime_serde")]
+    pub redelivery_interval: Option<Duration>,
```
Comment (0nkery) [inline src/config.rs:11]: Is it still configurable?

Comment (mgrachev) [reply src/config.rs]: No, thanks.

File: src/headers.rs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+#[derive(Debug, Clone)]
+pub struct Headers {
+    event_id: EventId,
+    sender_id: AgentId,
+}
+
+impl Headers {
+    pub(crate) fn new(event_id: EventId, sender_id: AgentId) -> Self {
```
Comment (manifest) [inline src/headers.rs:26]: Since we have optional properties for the struct, Builder pattern makes more sense.

**Review by 0nkery — APPROVED** [review]

File: src/headers.rs (hunk 9 above)
Comment (manifest) [reply src/headers.rs:30]: @mgrachev, just a reminder about this one.

Comment (mgrachev) [reply src/headers.rs:50]: Done

**Review by manifest — APPROVED** [review]

**Pull request merged**